api/graphql/types/recurring_reservation/serializers.py

```diff
@@ -437,3 +343,276 @@ def save(self, **kwargs: Any) -> RecurringReservation:
+
+        now = local_datetime()
+        first_reservation = self.instance.reservations.order_by("begin").first()
+        if begin_date != self.instance.begin_date and first_reservation.begin < now:
+            msg = (
+                "Reservation series' begin date cannot be changed after its "
+                "first reservation's start time is in the past."
+            )
+            raise ValidationError(msg, code=error_codes.RESERVATION_SERIES_ALREADY_STARTED)
+
+        validate_series_time_slots(
+            begin_date=begin_date,
+            begin_time=begin_time,
+            end_date=end_date,
+            end_time=end_time,
+            reservation_start_interval=interval,
+        )
+        return data
+
+    def save(self, **kwargs: Any) -> RecurringReservation:
+        skip_dates: list[datetime.date] = self.initial_data.get("skip_dates", [])
+        buffer_time_before: datetime.timedelta | None = self.initial_data.get("buffer_time_before")
+        buffer_time_after: datetime.timedelta | None = self.initial_data.get("buffer_time_after")
+
+        # Create both the recurring reservation and the reservations in a transaction.
+        # This way if we get, e.g., overlapping reservations, the whole operation is rolled back.
+        with transaction.atomic():
+            instance = super().save()
+            reservations = self.recreate_reservations(
+                instance,
+                buffer_time_before=buffer_time_before,
+                buffer_time_after=buffer_time_after,
+                skip_dates=skip_dates,
+            )
+
+        # Must refresh the materialized view since reservations changed time.
+        if settings.UPDATE_AFFECTING_TIME_SPANS:
+            update_affecting_time_spans_task.delay()
+
+        if settings.SAVE_RESERVATION_STATISTICS:
+            create_or_update_reservation_statistics.delay(
+                reservation_pks=[reservation.pk for reservation in reservations],
+            )
+
+        return instance
+
+    def recreate_reservations(
+        self,
+        instance: RecurringReservation,
+        buffer_time_before: datetime.timedelta | None,
+        buffer_time_after: datetime.timedelta | None,
+        skip_dates: list[datetime.date],
+    ) -> list[Reservation]:
+        now = local_datetime()
+        today = now.date()
+
+        # New reservations can overlap with existing reservations in this series
+        old_reservation_ids: list[int] = list(instance.reservations.values_list("pk", flat=True))
+
+        # Skip generating reservations for any dates where there is currently a cancelled or denied reservation.
+        # It's unlikely that the reserver will want or can have the same date even if the time is changed.
+        # Any exceptions can be handled after the fact.
+        skip_dates += list(
+            instance.reservations.exclude(
+                state__in=ReservationStateChoice.states_going_to_occur,
```

Comment:
Just in case, should exclude only `CONFIRMED` (i.e. skip everything else/anything unexpected) since future reservations of that state are what will be deleted below.

Reply:
Done.